first version of dlq handler

axthosarouris opened the pull request:
DLQ handler for using in Lambda destinations. Look at the comments for more information.

The PR is large due to the example Cloudformation templates.

axthosarouris commented on dlq/src/main/java/no/unit/nva/commons/dlq/DlqHandler.java:
```diff
@@ -0,0 +1,84 @@
+    @Override
+    public Void handleRequest(SQSEvent input, Context context) {
+
+        var failedEvents = input.getRecords().stream()
```
each DLQ event might contain multiple records

axthosarouris replied:
Probably yes. This handler however makes no assumptions for the SQS queue.6 If the queue has batch size 1 then each message will be handled separately.

anetteOlli replied:
I would prefer it if the failure of individual messages didn't cause the lambda to throw exception. Because if the batch always comes in the same order, we have a problem. Lets say message number 3 of 10 fails every time. Then messages 4 to 10 will never be processed. 
Isn't there supposed to be an object the lambda can return with list of unsuccessful messages?

But I think leaving it as it is now is okay, and then observe if we need to change it later is okay.

axthosarouris replied:
I will look into it.

> Isn't there supposed to be an object the lambda can return with list of unsuccessful messages?
I think this would be a DLQ for the DLQ :)

Bear in mind that we are talking about persisting a failed event, so a failure here would be failure to persist an already failed event. I can see the reason why we would like to persist as many failed events as possible, but at the same time, failing to do the one and only thing this handler is supposed to be doing should be an exception. Perhaps the solution would be to persist as many as and then throw an Exception. 
That being said, given that the handler is pushing Strings down to a firehose, failing to do so for one string would most likely mean that it would fail to do so for any subsequent string.

axthosarouris commented on dlq/src/test/resources/examples/dlq-stack.yaml:
```diff
@@ -0,0 +1,189 @@
+AWSTemplateFormatVersion: "2010-09-09"
```
Example DLQ stack that contains all resources and exposes only the DLQ ARN and the policy that allows lambdas to write to the DLQ. 
It can be deployed independently (i.e. one DLQ stack for the whole account) or as a nested stack (i.e. one nested DLQ stack for each deployed stack)

axthosarouris commented on dlq/src/test/resources/examples/use_dlq_stack.yaml:
```diff
@@ -0,0 +1,72 @@
+AWSTemplateFormatVersion: "2010-09-09"
```
Example of using the DLQ stack as nested stack

brinxmat commented on dlq/src/main/java/no/unit/nva/commons/dlq/DlqHandler.java:
```diff
@@ -0,0 +1,89 @@
+    public static final int NUMBER_OF_GROUPS = 10;
+    private final FirehoseClient firehoseClient;
+
+    public DlqHandler(FirehoseClient firehoseClient) {
```
There is no default constructor here

axthosarouris replied:
It is intentional. added documentation and a SampleDlqHandler in the tests

brinxmat commented on dlq/src/main/java/no/unit/nva/commons/dlq/DlqHandler.java:
```diff
@@ -0,0 +1,89 @@
+    }
+
+    @Override
+    public Void handleRequest(SQSEvent input, Context context) {
```
I suspect moving the logic to some class makes the code easier to comprehend.

```
package no.unit.nva.commons.dlq;

import com.google.common.collect.Lists;
import java.nio.charset.StandardCharsets;
import java.util.Collection;
import java.util.List;
import java.util.stream.Collectors;
import software.amazon.awssdk.core.SdkBytes;
import software.amazon.awssdk.services.firehose.FirehoseClient;
import software.amazon.awssdk.services.firehose.model.PutRecordBatchRequest;
import software.amazon.awssdk.services.firehose.model.Record;

public class FireHoseService {
    public static final int NUMBER_OF_GROUPS = 10;
    private final FirehoseClient client;
    private final String streamName;

    public FireHoseService(FirehoseClient client, String streamName) {
        this.client = client;
        this.streamName = streamName;
    }

    // API-wise, you might be tempted to return something other than Void, which can be used by the receiver.
    public Void persist(List<String> messages) {
        splitToSubLists(messages).forEach(this::pushToFirehose);
        return null;
    }

    private static Record createFirehoseRecord(String failedEvent) {
        return Record.builder().data(SdkBytes.fromString(failedEvent, StandardCharsets.UTF_8)).build();
    }

    private static Collection<List<String>> splitToSubLists(List<String> failedEvents) {
        return Lists.partition(failedEvents, NUMBER_OF_GROUPS);
    }

    private PutRecordBatchRequest assemblePutBatchRequest(List<Record> records) {
        return PutRecordBatchRequest.builder()
                   .records(records)
                   .deliveryStreamName(streamName)
                   .build();
    }

    private void pushToFirehose(List<String> group) {
        var records = group.stream()
                          .map(FireHoseService::createFirehoseRecord)
                          .collect(Collectors.toList());
        var request = assemblePutBatchRequest(records);
        client.putRecordBatch(request);
    }

```

brinxmat replied:
Also, it looks like the NUMBER:OF_GROUPS is a default value, which needs configuring.

axthosarouris replied:
Will do in next PR.

brinxmat commented on dlq/src/test/java/no/unit/nva/commons/dlq/DlqHandlerTest.java:
```diff
@@ -0,0 +1,86 @@
+    @ParameterizedTest(name = "number of messages:{1}")
+    @MethodSource("eventProvider")
+    void shouldPushFailedEventsToFirehoseForAutomaticallyStoringThemInS3AndOrganizeThemByTime(
+        Set<JsonNode> failedEvents, int numberOfEvents) {
```
`numberOfEvents` seems unused except in the test name, perhaps intentional, but a bit weird. If you want to affect changes in the name, probably (NB: you'll need to remove the name from `@ParameterizedTest`):

```
    public static Stream<Arguments> eventProvider() {
        return IntStream.of(1, 2, 10, 245).boxed()
                   .map(numberOfMessages ->
                        Arguments.of(Named.of(numberOfMessages +  " messages", randomJsons(numberOfMessages))));
    }
```

brinxmat commented on dlq/src/test/java/no/unit/nva/commons/dlq/DlqHandlerTest.java:
```diff
@@ -0,0 +1,86 @@
+
+        var dlqEvent = createDlqEventContainingFailedEvents(failedEvents);
+        handler.handleRequest(dlqEvent, EMPTY_CONTEXT);
+        var pushedRecords = firehoseClient.getRecords().stream()
+                                .map(record -> record.data().asUtf8String())
+                                .map(attempt(OBJECT_MAPPER::readTree))
+                                .map(Try::orElseThrow)
+                                .collect(Collectors.toSet());
```
Tempted to say that this is a suitably named method in FakeFirehoseClient

brinxmat replied:
I still think this :)

brinxmat commented on dlq/src/test/resources/examples/use_dlq_stack.yaml:
```diff
@@ -0,0 +1,72 @@
+    MemorySize: 1792
+    Tracing: Active
+    Timeout: 60
+    Runtime: java11
```
Still no Java17 for us :D

brinxmat commented on dlq/src/main/java/no/unit/nva/commons/dlq/PushToFirehoseService.java:
```diff
@@ -0,0 +1,48 @@
+
+public class PushToFirehoseService implements FailedEventHandlingService {
+
+    public static final Environment ENVIRONMENT = new Environment();
+    public static final String DELIVERY_STREAM_NAME = ENVIRONMENT.readEnv("DELIVERY_STREAM_NAME");
```
Can we pull these up to the handler? I feel like this is quite specific config, along with the batch size. I guess a config object can be injected to allow this.

axthosarouris commented:
I have removed all configuration and environment variables. There is no reason for them at this level, since the class needs to be subclassed to be used. The subclass will have all the configuration. Removed also the default methods because if we do not have any configuration, then we are just passing the same arguments ( fireshoseClient and firehose name) around.
